### ISSUE TITLE
DB lifecycle "beforeUpdate" does not contain data when draft & publish feature is turned on

### ISSUE THREAD

**Issue by M-erb** [strapi/strapi#8964]
### Describe the bug

To start the "beforeUpdate" lifecycle hook has two parameters (params and data). When a collection has the draft & publish feature is turned OFF the data parameter contains all of the collection entries direct columns/ fields. (ex. "name", "text" "created_by" and so on...) BUT when then draft & publish feature is turned ON this data parameter only ever has "published_at" property and none of the usual fields that are expected. This is happening weather the entry is published or in draft mode. As soon as I turn off the draft & publish feature on the collection the data parameter suddenly contains the fields from the entry.

Was trying to follow the "Create a slug system" guide in the documentation under "guides". But this is impossible when the draft & publish feature is turned on. [https://strapi.io/documentation/developer-docs/latest/guides/slug.html](https://strapi.io/documentation/developer-docs/latest/guides/slug.html)

### Steps to reproduce the behavior

My collections started off life with the draft & publish feature turned off so that is the order these steps take.
1. Create a collection with a field named "name" short text and make sure draft & publish under the "advanced" tab is turned OFF
2. Add a new entry on this collection with something added to "name"
3. Add "beforeUpdate" lifecycle hook that logs both parameters "params" and "data", example below.
4. Go back to your entry and edit it and save it.
5. Notice in the logs that the data parameter has the entries fields.
6. Goto "content-types builder" and edit this collection and turn ON the draft & publish feature under the "advanced" tab
7. Go back to your entry and edit it and save it(can be published or a draft, have not seen this change the data parameter in the lifecycle hook).
8. Notice in the logs that the data parameter does NOT have the fields it had before but only a "published_at" property.
9. Go back to "content-types builder" and edit this collection and turn OFF the draft & publish feature under the "advanced" tab
10. Repeat step 4
11. See step 5

### Expected behavior

That when a collection has the draft & publish feature turned on to still have the entries fields populate the "data" parameter on the "beforeUpdate" DB lifecycle hook. Currently I only see the entries fields on the data parameter when the draft & publish feature is turned off.

### Code snippets

```javascript
// Example of lifecycle hook code
module.exports = {
  lifecycles: {

    async beforeUpdate (params, data) {
      console.log('beforeUpdate params: ', params)
      console.log('beforeUpdate data: ', data)
    }

  }
}
```

### System

- Node.js version: v14.15.2
- NPM version: 6.14.9
- Strapi version: 3.4.0
- Database: MySql
- Operating system: Pop!_OS/Ubuntu

### Additional context

Was trying to follow the "Create a slug system" guide in the documentation under "guides". But this is impossible when the draft & publish feature is turned on. [https://strapi.io/documentation/developer-docs/latest/guides/slug.html](https://strapi.io/documentation/developer-docs/latest/guides/slug.html)

**Comment by derrickmehaffy** [comment]
Confirmed on v3.4.x

![image](https://user-images.githubusercontent.com/8593673/102832279-c7c0db00-43a2-11eb-8718-73addf1fb153.png)

**Comment by derrickmehaffy** [comment]
Current workaround:

- Take the ID from the params and make a direct query using `strapi.query('yourModel').findOne( { id: params.id } )` if the field `published_at` exists. Since when you publish, no new data is being saved, only the published_at field is updated.

**Comment by M-erb** [comment]
I appreciate this getting looked at so quickly @derrickmehaffy and also the work around.

I could be missing something though so just to be clear that this work around(aka querying the DB if "published_at" property is found) will actually give you the OLD data that is currently in the DB and NOT the new/being updated data. When the point of this lifecycle hook, in my mind, was to allow manipulating the new data(not in DB yet) before it was put into the DB. So basically this workaround will actually potentially introduce buggy/corrupted data and not allow a content editor to update something.

For example when wanting to auto populate a "URL" field for a blog post based on the title and maybe another field. Using the workaround above would not allow the URL to be properly updated until it is published which would be very confusing to a non-dev using the system. In our case all of our content editors are non-devs.

Please let me know if I am wrong on this point though.

**Comment by derrickmehaffy** [comment]
There are two actions happening here:

- You update content and save, this triggers the lifecycle with the new data (can't remember if it includes the old data, but I would imagine so since it's using a PUT request which within Strapi actually functions as a PATCH)
- You publish/unpublish the content

My example handles both of those use-cases as the `published_at` only exists when you publish or unpublish and thus makes the lookup, else if just uses the data that is already there.

**Comment by M-erb** [comment]
@derrickmehaffy, thank you again for looking into this. Testing this the "beforeUpdate" lifecycle hook contains the new data BEFORE it is put into the database. So with your example/workaround querying the database will give the OLD data and not the NEW data that is about to be put into the database.

Steps:
- Edit an existing entry(data is being changed)
- Lifecycle hook "beforeUpdate" is triggered and provides this new data
- Manipulate NEW data in "beforeUpate"
- "beforeUpdate" hook finishes and strapi enters NEW manipulated data into database.

If I were to query the database in this lifecycle hook then it would provide the OLD data(that is sitting in the database before save) and not the NEW data that is about to be saved to the database.

Basically this bug does not allow me to use both this "beforeUpdate" lifecycle hook to its fullest and also the draft & publish feature at the same time. Again because when draft & publish is turned on it only provides the "published_at" property when in draft mode and not the rest of the entries properties BEFORE they are put into the database.

Thank you again

**Comment by derrickmehaffy** [comment]
>  querying the database will give the OLD data

This is correct but, **only when you are publishing or unpublishing** an entry (when data isn't being changed) Because of how the Draft and publish functions you can't edit and also publish/unpublish at the same time.

So there are two different "actions" that are triggering the same lifecyle:

- Updating Data => Lifecycle uses new data
- Publish/Unpublish => There is no new data to use, so use what already exists.

Does that make sense?

**Comment by felixhagspiel** [comment]
I got the same issue. 

I got a blogpost collection with draft option turned on.

1. create a post via dashboard
2. publish post
3. edit post -> ´beforeUpdate´ hook returns `{ id: 15}` as `data` value
4. I run `const blogPost = await strapi.query('blog-post').findOne({id: data.id});` which returns the data from the database, **not** the edited data from the dashboard.

The best thing would be to send the edited data as `data` object instead of only sending the ID.

**Comment by felixhagspiel** [comment]
@M-erb I think I have found the solution.

`beforeUpdate ` provides 2 params, the second one is the edited data.

```
    async beforeUpdate(data, editedData) {
      console.info('update', data, editedData);
    },
```

**Comment by brunomptavares** [comment]
The same is happening to me but I have the draft/publish turned off for this content type. However I have a relation with a model which does, don't know if it's relevant.

Strange thing is if I print the "params" parameter it does have content inside, which actually looks like with what I was expecting from "data" parameter.

![issue](https://user-images.githubusercontent.com/15900346/116017752-62267900-a638-11eb-8272-855a0f223338.png)

**Comment by derrickmehaffy** [comment]
Fixed on v4

**Comment by panchicore** [comment]
as of today you have to make a couple of validations on `beforeUpdate` if you need to calculate something to save into the instance before the update, `publishedAt` won't be a property of `event.params.data` when user changes draft/publish option.

**Comment by pdhruv93** [comment]
v4 does not returns the currently stored relational data inside event object. It only sends the changes in the relational data using connect(array of ids added) and disconnect(array of ids removed). To get the current data you need to query DB.

```
const { data, where, select, populate } = event.params;
const id = where.id;
const existingData = await strapi.entityService.findOne("api::name.name", id, {
        populate: ["relationalField1", "relationalField2"],
      })
```
